File: src/main/java/at/bitfire/dav4android/property/ResourceType.java
package at.bitfire.dav4android.property;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;
import java.util.logging.Level;

import at.bitfire.dav4android.Constants;
import at.bitfire.dav4android.Property;
import at.bitfire.dav4android.PropertyFactory;
import at.bitfire.dav4android.XmlUtils;
import lombok.ToString;

@ToString
public class ResourceType implements Property {
    public static final Name NAME = new Name(XmlUtils.NS_WEBDAV, "resourcetype");

    public static final Name
            COLLECTION = new Name(XmlUtils.NS_WEBDAV, "collection"),    // WebDAV
            PRINCIPAL = new Name(XmlUtils.NS_WEBDAV, "principal"),      // WebDAV ACL
            ADDRESSBOOK = new Name(XmlUtils.NS_CARDDAV, "addressbook"), // CardDAV
            CALENDAR = new Name(XmlUtils.NS_CALDAV, "calendar");        // CalDAV

    public final Set<Property.Name> types = new HashSet<>();


    public static class Factory implements PropertyFactory {
        @Override
        public Name getName() {
            return NAME;
        }

        @Override
        public ResourceType create(XmlPullParser parser) {
            ResourceType type = new ResourceType();

            try {
                final int depth = parser.getDepth();

                int eventType = parser.getEventType();
                while (!(eventType == XmlPullParser.END_TAG && parser.getDepth() == depth)) {
                    if (eventType == XmlPullParser.START_TAG && parser.getDepth() == depth + 1) {
                        String namespace = parser.getNamespace(), name = parser.getName();

                        // use static objects to allow types.contains()
                        Name typeName = new Name(parser.getNamespace(), parser.getName());
                        if (COLLECTION.equals(typeName))
                            typeName = COLLECTION;
                        else if (PRINCIPAL.equals(typeName))
                            typeName = PRINCIPAL;
                        else if (ADDRESSBOOK.equals(typeName))
                            typeName = ADDRESSBOOK;
                        else if (CALENDAR.equals(typeName))
                            typeName = CALENDAR;

                        type.types.add(typeName);
                    }
                    eventType = parser.next();
                }
            } catch(XmlPullParserException|IOException e) {
                Constants.log.log(Level.SEVERE, "Couldn't parse <resourcetype>", e);
                return null;
            }

            return type;
        }
    }
}

File: src/main/java/at/bitfire/dav4android/property/ResourceType.kt
/*
 * Copyright © Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.dav4android.property

import at.bitfire.dav4android.Property
import at.bitfire.dav4android.PropertyFactory
import at.bitfire.dav4android.XmlUtils
import org.xmlpull.v1.XmlPullParser

class ResourceType: Property {

    companion object {
        @JvmField
        val NAME = Property.Name(XmlUtils.NS_WEBDAV, "resourcetype")

        @JvmField val COLLECTION = Property.Name(XmlUtils.NS_WEBDAV, "collection")    // WebDAV
        @JvmField val PRINCIPAL = Property.Name(XmlUtils.NS_WEBDAV, "principal")      // WebDAV ACL
        @JvmField val ADDRESSBOOK = Property.Name(XmlUtils.NS_CARDDAV, "addressbook") // CardDAV
        @JvmField val CALENDAR = Property.Name(XmlUtils.NS_CALDAV, "calendar")        // CalDAV
        @JvmField val SUBSCRIBED = Property.Name(XmlUtils.NS_CALENDARSERVER, "subscribed")
    }

    val types = mutableSetOf<Property.Name>()

    override fun toString() = "[${types.joinToString(", ")}]"


    class Factory: PropertyFactory {

        override fun getName() = NAME

        override fun create(parser: XmlPullParser): ResourceType? {
            val type = ResourceType()

            val depth = parser.depth
            var eventType = parser.eventType
            while (!(eventType == XmlPullParser.END_TAG && parser.depth == depth)) {
                if (eventType == XmlPullParser.START_TAG && parser.depth == depth + 1) {
                    // use static objects to allow types.contains()
                    var typeName = Property.Name(parser.namespace, parser.name)
                    when (typeName) {
                        COLLECTION -> typeName = COLLECTION
                        PRINCIPAL -> typeName = PRINCIPAL
                        ADDRESSBOOK -> typeName = ADDRESSBOOK
                        CALENDAR -> typeName = CALENDAR
                        SUBSCRIBED -> typeName = SUBSCRIBED
                    }
                    type.types.add(typeName)
                }
                eventType = parser.next()
            }

            return type
        }

    }

}
